go/pkg/client/cas.go

```diff
@@ -779,57 +779,81 @@ func marshalledRequestSize(d digest.Digest) int64 {
 
 // ReadBlob fetches a blob from the CAS into a byte slice.
 func (c *Client) ReadBlob(ctx context.Context, d digest.Digest) ([]byte, error) {
-	return c.readBlob(ctx, d.Hash, d.Size, 0, 0)
+	return c.readBlob(ctx, d, 0, 0)
 }
 
 // ReadBlobRange fetches a partial blob from the CAS into a byte slice, starting from offset bytes
 // and including at most limit bytes (or no limit if limit==0). The offset must be non-negative and
 // no greater than the size of the entire blob. The limit must not be negative, but offset+limit may
 // be greater than the size of the entire blob.
 func (c *Client) ReadBlobRange(ctx context.Context, d digest.Digest, offset, limit int64) ([]byte, error) {
-	return c.readBlob(ctx, d.Hash, d.Size, offset, limit)
+	return c.readBlob(ctx, d, offset, limit)
 }
 
-func (c *Client) readBlob(ctx context.Context, hash string, sizeBytes, offset, limit int64) ([]byte, error) {
+func (c *Client) readBlob(ctx context.Context, dg digest.Digest, offset, limit int64) ([]byte, error) {
 	// int might be 32-bit, in which case we could have a blob whose size is representable in int64
 	// but not int32, and thus can't fit in a slice. We can check for this by casting and seeing if
 	// the result is negative, since 32 bits is big enough wrap all out-of-range values of int64 to
 	// negative numbers. If int is 64-bits, the cast is a no-op and so the condition will always fail.
-	if int(sizeBytes) < 0 {
-		return nil, fmt.Errorf("digest size %d is too big to fit in a byte slice", sizeBytes)
+	if int(dg.Size) < 0 {
+		return nil, fmt.Errorf("digest size %d is too big to fit in a byte slice", dg.Size)
 	}
-	if offset > sizeBytes {
-		return nil, fmt.Errorf("offset %d out of range for a blob of size %d", offset, sizeBytes)
+	if offset > dg.Size {
+		return nil, fmt.Errorf("offset %d out of range for a blob of size %d", offset, dg.Size)
 	}
 	if offset < 0 {
 		return nil, fmt.Errorf("offset %d may not be negative", offset)
 	}
 	if limit < 0 {
 		return nil, fmt.Errorf("limit %d may not be negative", limit)
 	}
-	sz := sizeBytes - offset
+	sz := dg.Size - offset
 	if limit > 0 && limit < sz {
 		sz = limit
 	}
-	sz += bytes.MinRead // Pad size so bytes.Buffer does not reallocate.
-	buf := bytes.NewBuffer(make([]byte, 0, sz))
-	_, err := c.readBlobStreamed(ctx, hash, sizeBytes, offset, limit, buf)
-	return buf.Bytes(), err
+	// Pad size so bytes.Buffer does not reallocate.
+	buf := bytes.NewBuffer(make([]byte, 0, sz+bytes.MinRead))
+	// Compare the digest of the read blob to expected and retry the call on difference.
+	numRetries := 2
+	recvDg := digest.Empty
```

Comment:
better to use `var recvDg digest.Digest`?

Reply:
Sure.
Btw, I had a little bug in the logic which I now fixed, PTAL! Thank you!